[Plots] #638 New plot display options

akhenry opened the pull request:
Opening a new Pull Request for the merge into master. Suspect there will be a lot of comments to collect and the other PR is already quite long.

Note that this relates to issue #638 and has a previous PR associated with it at #645.

## Note:
There are some minor style changes that need to be made before this is committed, so ~~I have not included an author's checklist yet. Should not prevent code review from taking place though~~
@charlesh88 If you haven't started on [this](https://github.com/nasa/openmctweb/pull/645#issuecomment-180021486) I could have a crack at it anyway and have you review it as part of this PR?

## Changes
* Have defined a new attribute of Types called "regions". Regions are sections of the browse and edit screens for an object that, while fixed in location, have content that is completely customizable on a per-type basis. Regions are represented in code by a new Region class. Within Regions are RegionParts. A Region can have any number of RegionParts defined. The layout of both Regions and RegionParts are defined by representations. Whether a region part is visible within a region can depend on object state, and is determined by a policy of category region.

* Defined a new TypeRegionDecorator that defines standard regions for objects to avoid having to add region sections to all type definitions. Only have to explicitly define regions if a type has special requirements, such as the new plot display options.

* Defined a new EditableRegionPolicy which provides a convenience method for determining whether a region part should be visible in edit/browse mode based on a modes attribute in RegionPart definition.

* Modified the browse and edit screens to define an Inspector region. Currently this is the only region defined, however I will also defined 'object view' and 'elements' regions. Within the object inspector is the standard properties and location sections.

* Added a 'plot options' region part. Currently this is enabled on the Telemetry Plot type contained in the PlotOptions bundle (__not__ included by default).

* Made some modifications to mct-form to support listening for changes. Instead of using two-way data binding, I am listening for changes and triggering a mutation event on the domain object. This is so that there are discrete events that can be listened to from the plot side, rather than depending on Angular's data binding. Happy to discuss other options for this though.

larkin commented on example/plotOptions/bundle.js:
```diff
@@ -0,0 +1,145 @@
+                            }
+                        ]
+                    },
+                    "properties": [
```
Odd properties for the object-- is this necessary?

akhenry replied:
As above.

larkin commented on platform/commonUI/browse/res/templates/browse/object-properties.html:
```diff
@@ -0,0 +1,60 @@
+            <em class="t-inspector-part-header" title="The location of the original object that this was linked from.">Original Location</em>
+            <span class="inspector-location"
+                  ng-repeat="parent in primaryParents"
+                  ng-class="{ last:($index + 1) === primaryParents.length }">
```
@charlesh88 @akhenry Any reason why we use a class for "last" instead of the `last-child` selector? Removing this would reduce watches.

If the `last` class is required, we should at the least use the `$last` boolean provided by ng-repeat.

akhenry replied:
I don't honestly remember now whether I added this as part of my temporary CSS or if Charles added it. It's in a section of code that I don't think changed though, so it's probably been around for a while. From my end there's no need to have it, and I agree a CSS selector would be preferable here if it's technically feasible. 

@charles88 thoughts?

larkin replied:
I think it's been around for a while, just haven't noticed it until now.  It's a general pattern that we should avoid.

charlesh88 replied:
Can’t remember if I added this or not. We can use CSS last-child instead.

From: Pete Richards <notifications@github.com<mailto:notifications@github.com>>
Reply-To: nasa/openmctweb <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Monday, February 8, 2016 at 12:57 PM
To: nasa/openmctweb <openmctweb@noreply.github.com<mailto:openmctweb@noreply.github.com>>
Cc: Charles Hacskaylo <charles.f.hacskaylo@nasa.gov<mailto:charles.f.hacskaylo@nasa.gov>>
Subject: Re: [openmctweb] [Plots] #638 New plot display options (#651)


In platform/commonUI/browse/res/templates/browse/object-properties.html<https://github.com/nasa/openmctweb/pull/651#discussion_r52214295>:

> +            <span class="inspector-location"
> +                  ng-repeat="parent in contextutalParents"
> +                  ng-class="{ last:($index + 1) === contextualParents.length }">
> +                <mct-representation key="'label'"
> +                                    mct-object="parent"
> +                                    ng-model="ngModel"
> +                                    ng-click="ngModel.selectedObject = parent"
> +                                    class="location-item">
> +                </mct-representation>
> +            </span>
> +        </li>
> +        <li ng-if="primaryParents.length > 0">
> +            <em class="t-inspector-part-header" title="The location of the original object that this was linked from.">Original Location</em>
> +            <span class="inspector-location"
> +                  ng-repeat="parent in primaryParents"
> +                  ng-class="{ last:($index + 1) === primaryParents.length }">


@charlesh88<https://github.com/charlesh88> @akhenry<https://github.com/akhenry> Any reason why we use a class for "last" instead of the last-child selector? Removing this would reduce watches.

If the last class is required, we should at the least use the $last boolean provided by ng-repeat.

—
Reply to this email directly or view it on GitHub<https://github.com/nasa/openmctweb/pull/651/files#r52214295>.

larkin commented on platform/commonUI/browse/res/templates/browse/object-properties.html:
```diff
@@ -0,0 +1,60 @@
+            </span>
+        </li>
+    </ul>
+</div><!--/ holder-inspector -->
```
What does this comment mean?  I don't find a similar opening tag.

akhenry replied:
Will remove.

larkin commented on example/plotOptions/bundle.js:
```diff
@@ -0,0 +1,145 @@
+                    "regions": {
+                        "inspector": plotInspector
+                    },
+                    "telemetry": {
```
Probably don't need all this config from sine waves here.

akhenry replied:
Originally had a blank plot, but added the sine wave config so that the plot shows something. Blank plot would be ok if telemetry could be dropped onto it, but currently that would result in a new panel which is not configured for new plot options.

Can remove sine wave stuff if a blank plot is preferable here?

larkin replied:
Ah, I would expect that one would "create" this object and then drag a telemetry point into this object.  Guess the trick here is the drag gesture is a little weird with plot functionality.  If you were to trigger editing before starting the drag gesture it would work properly.  For now this is fine, it's an example bundle; we'll clean it up when we finish this work here.

akhenry replied:
Agree, what you're describing is the ideal behavior. Wanted to limit scope of this change and not have to refactor drop handling. This was only ever intended as a temporary measure as I assumed you were creating a new object type for your plot work? If this is not the case, then I can do this as part of the integration step.

larkin replied:
Yeah don't change anything regarding this, post integration this bundle may no longer exist.

larkin commented on platform/commonUI/browse/res/templates/browse/inspector-region.html:
```diff
@@ -0,0 +1,30 @@
+                ng-model="ngModel">
+        </mct-representation>
+    </div>
+</div><!--/ PaneController -->
```
Panecontroller?

akhenry replied:
Copy-paste detritus. Will remove.

larkin commented on platform/commonUI/browse/res/templates/browse/object-properties.html:
```diff
@@ -0,0 +1,60 @@
+            <em class="t-inspector-part-header">Properties</em>
+            <div class="inspector-properties"
+                 ng-repeat="data in metadata"
+                 ng-class="{ first:$index === 0 }">
```
Can use $first instead of index comparison.  Also; `:first-child` selector would be ideal and reduce watch counts for something that CSS should be handling.

akhenry replied:
Agree, CSS would be preferable for this. @charlesh88 I could make this change, but also happy for you to handle it if you'd prefer to.

larkin commented on platform/commonUI/browse/res/templates/browse/object-properties.html:
```diff
@@ -0,0 +1,60 @@
+            <em class="t-inspector-part-header" title="The location of this linked object.">Location</em>
+            <span class="inspector-location"
+                  ng-repeat="parent in contextutalParents"
+                  ng-class="{ last:($index + 1) === contextualParents.length }">
```
same as elsewhere: use `$last` instead of `$index`, ideally use `:last-child` in CSS instead of adding a watch via ng-class.

larkin commented on platform/features/plot/src/PlotOptionsController.js:
```diff
@@ -0,0 +1,145 @@
+         * @constructor
+         * @param {Scope} $scope the controller's Angular scope
+         */
+        function PlotOptionsController($scope, topic) {
```
Style wise, this class should be prototype based.

akhenry replied:
Quite right, I'll change it.

larkin commented on platform/features/plot/src/PlotOptionsController.js:
```diff
@@ -0,0 +1,145 @@
+                configuration.plot = configuration.plot || {};
+                configuration.plot.xAxis = configuration.plot.xAxis || {};
+                configuration.plot.yAxis = configuration.plot.yAxis || {}; // y-axes will be associative array keyed on axis key
+                configuration.plot.series = configuration.plot.series || {}; // series will be associative array keyed on sub-object id
```
Wasn't this to be an array of objects to allow for the potential case of adding an object more than once?  Becomes harder to make this change later if we don't do this up front (have to migrate existing configurations).

akhenry replied:
As per https://github.com/nasa/openmctweb/pull/645#issuecomment-180021486 I was hoping to leave this for the "wiring up" stage with the actual plot. It complicates things in the forms a bit, and wanted to get this in before code freeze to spread out the code reviews a little. Let me see if I can get something in today.

larkin commented on platform/features/plot/src/PlotOptionsForm.js:
```diff
@@ -0,0 +1,166 @@
+            var self = this;
+            this.onchangeTopic = topic();
+
+            function onchange(value){
```
Hrm, it seems crazy that we have to explicitly listen to every field to bubble up an on-change event.  Is it possible for the mct-controls to $emit change events so that change events bubble up to a singular listener?  Or a similar solution?

VWoeltjen replied:
I think this could be a controller responsibility; controller could `$watchCollection` on the form models, removing the need for `PlotOptionsForm.listen` entirely. Controller looks to be the only consumer of that method, in any case.

akhenry replied:
It's a *bit* crazy to specify it against each control in the form spec for this particular case, where you're only interested in a single notification for any change in the form values, I agree. 

What I think is not crazy is having the __ability__ to have per-control change handling, and extending the form specification standard to support this.

The rationale behind this was that I wanted to avoid two-way data binding and additional watches, etc, and to implement a _somewhat_ angular agnostic approach. Granted, ng-change is an angular directive, and behaves differently to onchange, but it's similar enough that I think ng-change could be refactored out in future for an approach that uses native event handling, or the event handling of whatever view framework we end up using.

That being said, I'm open to the idea of extending mct-form to listen to changes in all its controls automatically, to allow for a single onchange/ng-change event handler to be registered.

larkin replied:
Absolutely agree: should be able to do per-control change handling, at the same time, there should be a "bubble up" change handler for a given form so that it is not necessary to explicitly listen for changes on every field when you want "all fields in this form".

larkin commented on platform/forms/res/templates/form.html:
```diff
@@ -19,35 +19,29 @@
             {{section.name}}
         </div>
-        <div class="form-section">
+        <div class="form-section" ng-class="{ first:$index < 1 }">
```
better to use `$first` or better yet `:first-child` (e.g. `.section:first-child > .form-section { /*css*/};`).

larkin commented:
Architecturally, a few thoughts.  I like Regions but we need to ensure the needs of our current situation don't result in us designing something that won't work in the future.

In the context of this pull request, I would scope the inspector overrides in such a way that they are very specific to the use of the inspector and they don't assume too much about how we might handle region management as an extension point in the future.

Some background thoughts about regions:

* A Region is a single location that can display any View.
* Views define their regions, for instance, the browse view would define `tree`, `toolbar`, `main`, `bottombar`, and `inspector` regions, among others.
* have some standard for passing configuration to sub views that makes view flow very easy.
* allow nesting of regions with intermediate views that themselves define configuration.  So a view for the `inspector` region might define `metadata` and `elements` regions.
* Allow path based locating of views, such that a developer could override the `inspector` region, or they could override the `inspector.metadata` region.

Which brings us to RegionParts.  They assume a region contains n views: not only do they show a single view in a region, they could potentially show multiple views in a region.  This conflates the concerns of a Region (show this view here) with the specific implementation necessary for the Inspector (display n sections inside this view).

I would prefer to stick to a "one view per region" standard, as this allows path based view replacement, e.g. render "MyView" at region "inspector.metadata" based on "some attribute of a domain object".

For the inspector, there would be two options: either consider it a view which renders n sections based on policy (as currently implemented) or have it define regions ('metadata', 'form') and have developers specify a custom view which is peer to the metadata, while allowing them to override the metadata view and disable it.

@akhenry @vwoeltjen interested in hearing your thoughts.

akhenry commented:
@larkin I take your point. RegionParts were designed specifically to meet the needs of the inspector and to allow policies to determine visibility of individual sections within the region. It's not necessarily a bad thing that they were designed to meet an immediate need though, and it's been done in a generic way that is extensible. I think there are arguments for and against doing it this way, and I don't feel entirely comfortable with the additional API complexity that `RegionPart`s add. 

> have some standard for passing configuration to sub views that makes view flow very easy

I'm not sure what this means, could you clarify? Trying to minimize configuration, and instead rely on templates and programmatic interfaces.

I like the idea of having addressable regions. I think a lot of the concerns you've raised here can be addressed with a semantic change, and some minor changes to the API. What I propose is removing the concept of RegionParts and allow `Region`s to contain `Region`s within them that are addressable via a canonical path (`inspector.metadata` etc.). Within the markup, you could either enumerate all subregions without caring what they are (as currently happens with RegionParts) or you would have the option of retrieving them individually using their canonical path where special markup is required.

VWoeltjen commented:
> In the context of this pull request, I would scope the inspector overrides in such a way that they are very specific to the use of the inspector and they don't assume too much about how we might handle region management as an extension point in the future.

Agree.

> I would prefer to stick to a "one view per region" standard, as this allows path based view replacement, e.g. render "MyView" at region "inspector.metadata" based on "some attribute of a domain object".

Agree, but think we should wait for #463 to really settle on this. Instead of a Region with _n_ RegionParts, would prefer just _n_ Regions for cases like this.

For adding Inspector content (e.g. new sections) via plugin, I see supporting one or both following options in the post-refactor API:

* Expose the run-time instance of the Inspector through a service, allowing it to be decorated (in this case, using this decoration to add more regions)
* Just have an `defineRegion(title : string, populateFn : (function (domainObject : DomainObject) : View)` method on Inspector, allowing plugins to create and use new regions within it.
  * This would be useful to DRYly ensure a consistent look and feel among Inspector regions, along with the normal replace-when-the-object-changes behavior.
  * Inspector (or some superclass that it uses) should be smart enough to hide regions when empty.
  * Method signature could be tweaked a bit (e.g. add a `key` to allow named regions to be _overridden_)

That said, I think this is more like a terminology problem than an actual implementation problem in the context of this issue. In the current implementation, I think we could change `Region` to `RegionGroup` and `RegionPart` to `Region` and be basically compatible with the way we intend to use the term in the future.

larkin commented:
Right-- I think this is a terminology problem in the context of this issue.  I'd rather stick this under a new category (not call the type attribute "regions") with a clear implementation that directly impacts the inspector without any intermediate concepts.  We would then use this as a design input when looking at Regions-- which could become a very big topic that looks very different than what we are currently looking for.

akhenry commented:
@larkin Hrm, this has some impact on #627 as my current implementation in that branch is based on the Regions API implemented here. Basically in addition to the Inspector region I introduce a 'View' region (although I prefer your use of 'Main' here so I'm inclined to use that instead) which uses policy control to hide/show edit/browse views depending on object edit state. 

What you're saying here implies a major rewrite of Regions later, in which case it doesn't make much sense to extend its use anywhere else beyond the Inspector for now. That being said, regions makes #627 a lot easier and I'm disinclined to take a different approach for that issue that addresses a similar issue (customizing views based on object state).

larkin commented:
If we want to reuse for #627 then want to talk a bit about terminology & API: what will be implemented in V1.0, and is that the final version of that going to look like (#463).  So long as we have a clear pathway to refactor going forward (with no persistence migrations), then I'm happy.

larkin commented:
@akhenry Can you put an author checklist on this and assign for a formal review & integrate?

akhenry commented:
@larkin I most certainly can.

It looks as though this is not going to make it in before code freeze. That's not vitally important as it would not be tested as part of the testathon anyway.

### Author Checklist

1. Changes address original issue? __Y__
2. Unit tests included and/or updated with changes? __Y__
3. Command line build passes? __Y__
4. Changes have been smoke-tested? __Y__

akhenry commented:
@larkin Have added author checklist, but have left assignment to me for now as I am addressing the issues raised in your code review.

akhenry commented:
> I'd rather stick this under a new category (not call the type attribute "regions") with a clear implementation that directly impacts the inspector without any intermediate concepts. 

Ok, to address the concerns here I could:
* Call the type attribute "inspector".
* Do away with `RegionPart`s, and refactor their attributes into `Region`s themselves.
* Refactor regions to be constituted of sub-`Region`s instead of `RegionPart`s

This would be a fairly minimal amount of work, and would address a lot of the issues raised here I think.

akhenry commented:
I can see there being concerns about the use of sub-regions in the suggestion above, so an alternative would be to do away with sub-regions and just define three different inspectors for -
* plots in browse mode
* plots in edit mode * still not sure there are differences here yet
* everything else

Use a policy to decide which to show.

This might actually take a little longer than implementing sub-regions, and might result in a bunch of duplicate markup, but we don't go as far down the road of defining a Regions API.

akhenry commented:
Changes

- Removed TypeRegionDecorator
- Removed RegionParts
- Added an onchange ability to forms, no longer necessary to register change listener on all controls
- Changed type attribute to Inspector

Still to do
- Style changes (use last-child etc.) - vote for deferring this, pre-existing problem
- form events - Still some discussion to be had around this. Have extended mcc-form to support form level change events. Missed @vwoeltjen ’s [comment](https://github.com/nasa/openmctweb/pull/651#discussion_r52227390) re: watchCollection (apologies). Did consider this, but was trying to avoid watches as much as possible and make things non-angular-ish. I'm not that married to it though, so I'd be quite happy to go with your suggested approach and minimize the changes to mct-form.

akhenry commented:
Ok, so I've switched to a $watchCollection based approach now, and removed all listen capabilities from the forms and controllers.

It's ready to re-review.

larkin commented:
@akhenry I resolved merge conflicts here, but there are failing tests due to 7d0de68.  Can you fix those and reassign to me when complete?

akhenry commented:
@larkin fixed the failing tests, all yours

akhenry commented:
New PR #708 for new table component makes use of region functionality provided by this PR. Branch Open627_rebase also makes use of this region functionality and will wait until this is merged to create a PR for that branch.

larkin commented:
# Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
